### ISSUE TITLE
N^2 problem when registering custom classes.

### ISSUE THREAD

**Issue by robertnishihara** [ray-project/ray#1002]
Whenever a custom class is serialized by Ray for the first time on a worker (or driver), the class definition is pickled and shipped to all of the other workers (and the driver).

So if the same class definition is serialized on all of the workers, it will be exported to all of the other workers every single time, resulting in `N^2` messages, where `N` is the number of workers. This can be inspected as follows (on a single machine).

```python
import ray
import time

N = 10

ray.init(num_cpus=N)

class Foo(object):
    pass

@ray.remote
def f():
    time.sleep(0.1)
    return Foo()

time.sleep(3)

for _ in range(N):
    f.remote()

time.sleep(5)

# Then see how many things have been "exported".
exports = ray.worker.global_worker.redis_client.lrange('Exports', 0, -1)

print(len(exports))  # This should be something like N + 6 (assuming 
                     # that all N tasks got scheduled on different workers).
```

When the number of workers is in the thousands (e.g., 5000 across a cluster), this seems to congest the primary Redis shard a lot (leading to other kinds of failures).

**Comment by robertnishihara** [comment]
A natural solution here is to use a hash of the string we get by pickling the class to identify the class and therefore deduplicate exported classes. Each worker would still try to export the class, but only one of them would succeed and so each worker would only import the class once.

I remember thinking that the string produced when cloudpickling a class was nondeterministic, but now I'm having trouble reproducing the nondeterminism.

**Comment by robertnishihara** [comment]
Fixed by #1038.
